### PR TITLE
Controls nullability cleanup

### PR DESCRIPTION
**Pull request by MrJul**
Another pass of nullability fixes, this time for the `Controls.*` assemblies.

Notable changes:
 - Various `Panel` descendants were checking for potential `null` children. These checks have been removed since `Controls` is a list of `Control`, not `Control?`. (If necessary, a null check can be added to `Controls.Add` – imo nullability annotations are sufficient and should be trusted).
 - `Grid`: most methods assume that `ExtendedData` isn't null or that some members of this class aren't null at some specific points (`DefinitionsU/V`, `TempDefinitions`...). A few `MemberNotNull` attributes have been added when possible but it's still quite hard to follow the whole flow. Those methods should ideally all be instance methods of `ExtendedData` with proper annotations, but that's a refactoring way too large for this PR.

### CODE REVIEW

**Comment by avaloniaui-team** [comment]
You can test this PR using the following package version. `11.0.999-cibuild0029719-beta`. (feed url: https://pkgs.dev.azure.com/AvaloniaUI/AvaloniaUI/_packaging/avalonia-all/nuget/v3/index.json) [PRBUILDID]

File: src/Avalonia.Controls/TreeView.cs [hunk 1]
```diff
@@ -132,6 +128,7 @@ public SelectionMode SelectionMode
         /// <summary>
         /// Gets or sets the selected items.
         /// </summary>
+        [AllowNull]
         public IList SelectedItems
```
Comment (maxkatz6) [inline src/Avalonia.Controls/TreeView.cs:132]: Why is attribute used here instead of "?" ?

File: src/Avalonia.Controls/Primitives/SelectingItemsControl.cs [hunk 2]
```diff
@@ -551,6 +558,7 @@ bool Match(Control container)
             base.OnTextInput(e);
         }
 
+        /// <inheritdoc />
```
Comment (maxkatz6) [inline src/Avalonia.Controls/Primitives/SelectingItemsControl.cs:561]: XML docs for couple of years is now automatically inherited. inheritdoc is not necessary here.

File: src/Avalonia.Controls/Primitives/TemplatedControl.cs [hunk 3]
```diff
@@ -290,12 +290,6 @@ public sealed override void ApplyTemplate()
                     ApplyTemplatedParent(child, this);
                     ((ISetLogicalParent)child).SetParent(this);
                     VisualChildren.Add(child);
-                    
-                    // Existing code kinda expect to see a NameScope even if it's empty
-                    if (nameScope == null)
-                    {
-                        nameScope = new NameScope();
-                    }
```
Comment (maxkatz6) [inline src/Avalonia.Controls/Primitives/TemplatedControl.cs:299]: I am unsure if namescope won't be ever null here.
Custom template technically could return null as a namescope. Unless we rely on nullability attributes in the new user's code, and assume that it's rare to have custom templates.

**Comment by maxkatz6** [comment]
> Various Panel descendants were checking for potential null children

IIRC, currently it fails with NRE somewhere in AvaloniaList, when developer adds a null child.
It really should throw more specific error in case of Panel.Children. If not in this PR, then we need an issue for later.

File: src/Avalonia.Controls/Documents/Span.cs [hunk 4]
```diff
@@ -68,26 +70,26 @@ internal override void OnInlineHostChanged(IInlineHost? oldValue, IInlineHost? n
...
-                oldValue.Invalidated -= (s, e) => InlineHost?.Invalidate();
+                oldValue.Invalidated -= OnInlinesInvalidated;
             }
 
             if (newValue is not null)
             {
                 newValue.LogicalChildren = LogicalChildren;
                 newValue.InlineHost = InlineHost;
-                newValue.Invalidated += (s, e) => InlineHost?.Invalidate();
+                newValue.Invalidated += OnInlinesInvalidated;
```
Comment (maxkatz6) [inline src/Avalonia.Controls/Documents/Span.cs:92]: Feels like it was a bug before.

File: src/Avalonia.Controls/Grid.cs [hunk 5]
```diff
@@ -164,41 +164,43 @@ public bool ShowGridLines
         /// <summary>
         /// Returns a ColumnDefinitions of column definitions.
         /// </summary>
+        [MemberNotNull(nameof(_extData))]
```
Comment (maxkatz6) [inline src/Avalonia.Controls/Grid.cs:167]: Does this attribute work here? I thought it makes sense only when set on specific get and set methods inside.

File: src/Avalonia.Controls/TreeView.cs (hunk 1 above)
Comment (MrJul) [reply src/Avalonia.Controls/TreeView.cs:132]: To match the existing accessor implementations. The getter should imo definitely stay not null as it's very convenient: callers can add/remove items to the collection directly. It's also the backing property for the `SelectedItemsProperty`. The setter already accepted null and I reflected that by using `AllowNull` to annotate  the setter without affecting the getter.

Comment (maxkatz6) [reply src/Avalonia.Controls/TreeView.cs:132]: So, I suppose setting AllowNull on the whole property affects only setter, while not affecting getter. Right?
Like an equivalent of `{ get {} [AllowNull] set {} }`

File: src/Avalonia.Controls/Primitives/SelectingItemsControl.cs (hunk 2 above)
Comment (MrJul) [reply src/Avalonia.Controls/Primitives/SelectingItemsControl.cs:561]: I thought it still triggered the compiler warning about missing the XML documentation (as a suggestion here). I can revert those if you prefer, I added it to all methods which had inherited doc on files I modified, might have been a little carried away here :)

Comment (maxkatz6) [reply src/Avalonia.Controls/Primitives/SelectingItemsControl.cs:561]: Needs to be double checked if we have these warnings by not having inheritdoc here.
Good question if we should add inheritdoc then, or disable warning and report to the https://github.com/dotnet/roslyn-analyzers ...

File: src/Avalonia.Controls/TreeView.cs (hunk 1 above)
Comment (MrJul) [reply src/Avalonia.Controls/TreeView.cs:132]: Yes, `AllowNull` isn't valid on methods and only affects input, so the setter. Not a very common pattern, but you can find it in [some places in the runtime](https://github.com/dotnet/runtime/blob/d913b94d041e192ab2f389ecae6b972da2094560/src/libraries/System.Diagnostics.Process/src/System/Diagnostics/ProcessStartInfo.cs#L61). `MaybeNull` only affects the getter.

File: src/Avalonia.Controls/Documents/Span.cs (hunk 4 above)
Comment (MrJul) [reply src/Avalonia.Controls/Documents/Span.cs:92]: Definitely, the `-=` above didn't do anything.

File: src/Avalonia.Controls/Primitives/TemplatedControl.cs (hunk 3 above)
Comment (MrJul) [reply src/Avalonia.Controls/Primitives/TemplatedControl.cs:299]: `ITemplateResult.NameScope` was already correctly annotated as non-null. Everyone is still free to ignore any annotation, but shouldn't complain too much when things break imo :)

File: src/Avalonia.Controls/Grid.cs (hunk 5 above)
Comment (MrJul) [reply src/Avalonia.Controls/Grid.cs:167]: Yes, it has the same effect as putting the attribute on both the get and set methods.

**Comment by MrJul** [comment]
> It really should throw more specific error in case of Panel.Children. If not in this PR, then we need an issue for later.

I can add that, the `Controls` class seems like a good place, setting `base.Validate`. The only issue is that `Validate` can be changed by any consumer since it's public. Ideally the real list should be encapsulated, like in `Styles`. However there are precedents for simply deriving from `AvaloniaList` and changing `Validate`: `Transitions`, `DataTemplates`.

File: src/Avalonia.Controls/Primitives/SelectingItemsControl.cs (hunk 2 above)
Comment (MrJul) [reply src/Avalonia.Controls/Primitives/SelectingItemsControl.cs:561]: I double checked: `inheritdoc` is definitely needed to prevent the warning (or the suggestion here since the severity has been changed in `.editorconfig`.)

https://github.com/dotnet/roslyn/issues/65386#issuecomment-1312324309 confirms that documentation comments aren't automatically inherited (they don't appear in the resulting XML file without `inheritdoc`): it's an IDE-time feature for convenience.

**Comment by MrJul** [comment]
Added a null item check and and associated test for `Controls`. The `Controls` constructor taking an `IEnumerable<Control>` has a base virtual call to `AddRange`, which I generally prefer to avoid but it won't cause any problems here.

**Comment by avaloniaui-team** [comment]
You can test this PR using the following package version. `11.0.999-cibuild0029737-beta`. (feed url: https://pkgs.dev.azure.com/AvaloniaUI/AvaloniaUI/_packaging/avalonia-all/nuget/v3/index.json) [PRBUILDID]

**Review by maxkatz6 — APPROVED** [review]

**Comment by avaloniaui-team** [comment]
You can test this PR using the following package version. `11.0.999-cibuild0029770-beta`. (feed url: https://pkgs.dev.azure.com/AvaloniaUI/AvaloniaUI/_packaging/avalonia-all/nuget/v3/index.json) [PRBUILDID]
